### PR TITLE
Verification options for TLS 

### PR DESCRIPTION
**Pull request by weargoggles**
This is my attempt at #2573 

I don't know where to put the defaults?

### CODE REVIEW

**Comment by weargoggles** [comment]
I find it completely impossible to understand the output of the test suite for this failure. Can someone help me?

**Comment by weargoggles** [comment]
Hi @schmichael if you find some time can you help me understand why the test suite is failing for me, and what I might have got wrong?

**Comment by schmichael** [comment]
@weargoggles Thanks for the PR! You made my week easier. :)

The only failure I see is `TestHTTP_Stream_Truncate` which I think is just a flaky test. I'd ok merging with that failing.

The spirit of your PR is perfect and should work as intended, but we'd actually like to simplify the configuration a little bit:

Instead of exposing `verify_{incoming,outgoing}` only have a single configuration parameter: `verify_https_client`

That parameter should default to `false` and control `VerifyIncoming` on the HTTP server's TLS config.

When `tls.rpc=true` we shouldn't even allow people to skip `Verify{Incoming,Outgoing}`. With `tls.http=true` we still want to allow verifying client certificates to be optional, hence the new `verify_https_client`.

Sorry for changing the goal on you and thanks again for the PR!

**Comment by weargoggles** [comment]
@schmichael I've got a sort of secondary issue, which is that I have trouble health-checking nomad's HTTP API when HTTPS is enabled. My load balancer can't have a client certificate. So is there room for a secondary HTTPS port, à la Consul?

File: command/agent/http.go [hunk 1]
```diff
@@ -65,8 +65,8 @@ func NewHTTPServer(agent *Agent, config *Config) (*HTTPServer, error) {
 	// If TLS is enabled, wrap the listener with a TLS listener
 	if config.TLSConfig.EnableHTTP {
 		tlsConf := &tlsutil.Config{
-			VerifyIncoming:       false,
-			VerifyOutgoing:       true,
+			VerifyIncoming:       config.TLSConfig.VerifyHTTPSClient,
+			VerifyOutgoing:       config.TLSConfig.VerifyHTTPSClient,
```
Comment (schmichael) [inline command/agent/http.go]: This should still be hard-coded as `true`. If `TLSConfig.EnableHTTP` is set we don't want to even allow skipping server certificate verification.

**Comment by schmichael** [comment]
> I've got a sort of secondary issue, which is that I have trouble health-checking nomad's HTTP API when HTTPS is enabled. My load balancer can't have a client certificate. So is there room for an optional HTTPS port, like Consul?

Unfortunately, I think we're going to wait until we rework our networking code (binding, advertisement, IP pools, etc) to support HTTP and HTTPS. Allowing both obviously negates the benefits of your change without external firewalling. Any chance your 

You did remind me Consul's Nomad HTTP health checks will have this same issue, so we need to switch them to a `tcp` check if `VerifyHTTPSClient==true`: https://github.com/hashicorp/nomad/blob/master/command/agent/agent.go#L382-L383

I can make that change after merging this PR if you're not a Consul user.

**Comment by weargoggles** [comment]
@schmichael I'll explain a bit more clearly. My load balancer (AWS ELB) can run its health check with a different port and protocol to the one it uses as the backend. So:
- Consul is configured to have both HTTP and HTTPS ports
- The load balancer is just speaking TCP for the incoming connections
- the health check is HTTP, and the firewall is configured so that only the load balancer can access Consul's insecure HTTP port

I want to be able to do the same thing in Nomad; forward TCP connections (leaving TLS to Nomad) based on the health determined by a plain HTTP check.

I am using Consul to bootstrap and advertise. The configuration in the Consul health check you mention works perfectly for me, and it would work even if the 'skip verification' option was not set, because the certificates I'm using are valid. Perhaps that should be configurable; but I don't know that it can be controlled by `verify_https_client` because it's Consul being configured there.

**Comment by weargoggles** [comment]
@schmichael Do you consider your feedback addressed? My test failure doesn't look related to this change.

**Comment by schmichael** [comment]
Merged. Thanks @weargoggles!

Created #2606 to track the other feature.

**Comment by shilov** [comment]
This deserves its own issue but I'll mention it here in the interim.. I had Consul configured with a self-signed cert and Nomad was able to register services, but health checks would fail due to SSL verification. The `verify_ssl` flag was false, which is what must've allowed the service registration to succeed, but the health checks threw an error about x509 authority being unknown.

I'll try to make a separate issue about this later.

**Comment by weargoggles** [comment]
@shilov Consul supports `TLSSkipVerify` on HTTPS checks but Nomad doesn't yet expose this option in service stanzas.

**Comment by schmichael** [comment]
> @shilov Consul supports TLSSkipVerify on HTTPS checks but Nomad doesn't yet expose this option in service stanzas.

Fixed in master!

**Comment by shilov** [comment]
Kudos! 👍

**Pull request merged**